File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.cpp
class MedianFinder {
public:
    priority_queue<int>mx;
    priority_queue<int, vector<int>, greater<int>>mn;
    MedianFinder() {
        
    }
    
    void addNum(int num) {
        if(mx.empty()) {
            mx.push(num);
        }
        else if(mx.size() == mn.size()) {
            int x = mn.top(); mn.pop();
            if(num > x) {
                mn.push(num);
                mx.push(x);
            }
            else {
                mn.push(x);
                mx.push(num);
            }
        }
        else {
            int x = mx.top(); mx.pop();
            if(num < x) {
                mx.push(num);
                mn.push(x);
            }
            else {
                mn.push(num);
                mx.push(x);
            }
        }
    }
    
    double findMedian() {
        if((mx.size() + mn.size())%2 == 1) {
            return (double)mx.top();
        }
        else {
            return (double) (mn.top() + mx.top())/2;
        }
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */

File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.java
class MedianFinder {
    PriorityQueue<Integer>mn;
    PriorityQueue<Integer>mx;
    public MedianFinder() {
        mn = new PriorityQueue();
        mx = new PriorityQueue(Collections.reverseOrder());
    }
    
    public void addNum(int num) {
        if(mx.isEmpty()) {
            mx.offer(num);
        }
        else if(mx.size() == mn.size()) {
            if(num > mn.peek()) {
                int x = mn.poll();
                mx.offer(x);
                mn.offer(num);
            }
            else {
                mx.offer(num);
            }
        }
        else {
            if(num < mx.peek()) {
                int x = mx.poll();
                mn.offer(x);
                mx.offer(num);
            }
            else {
                mn.offer(num);
            }
        }
    }
    
    public double findMedian() {
        if(mx.size() == mn.size()) {
            return (double) (mx.peek() + mn.peek())/2;
        }
        else {
            return (double) mx.peek();
        }
    }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */